### PR TITLE
[DDP][BE] Remove bucket replicas

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#73237 [DDP][BE] Remove bucket replicas**

Differential Revision: [D34399176](https://our.internmc.facebook.com/intern/diff/D34399176)

**Overview**
Fixes https://github.com/pytorch/pytorch/issues/72539.

I opted to move all members of `BucketReplica` directly into `Bucket` since I did not see any logical separation that justified having a nested `struct`.

**Test Plan**
I was seeing some flaky tests when running through the distributed test suite using A100s (e.g. `test_broadcast_coalesced_nccl()`). However, these tests should have no dependency on the contents of the modified files (i.e. `reducer.cpp`, `reducer.hpp`, etc.), so I suspect they may be related to my specific environment.

Since the reducer-related tests were passing locally, I opted to just confirm the rest using CI.

### CODE REVIEW

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34399176).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34399176).

File: torch/csrc/distributed/c10d/reducer.hpp [hunk 1]
```diff
@@ -344,117 +345,104 @@ class TORCH_API Reducer {
...
+  // and/or faster time to completion. Coalescing requires the constituent
+  // gradients to have the same dtype and device, and the resulting flattened
+  // tensor uses that common dtype and device. The bucket's contents are filled
+  // as their corresponding gradients are computed (triggered by autograd
+  // hooks), and the buckets are reduced in a predetermined order consistent
+  // across processes.
   struct Bucket {
-    std::vector<BucketReplica> replicas;
+    // Flattened (1-dimensional) contents of the bucket
+    at::Tensor contents;
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.hpp:380]: These are gradients, should we rename them as such?

File: torch/csrc/distributed/c10d/reducer.hpp [hunk 2]
```diff
@@ -344,117 +345,104 @@ class TORCH_API Reducer {
...
-  // structure. The `bucket_index` field points to the bucket in the `buckets_`
-  // vector. The `intra_bucket_index` field points to the index of the variable
-  // in any of the vector fields in the bucket replica.
+  // A variable locator locates a particular variable in the reducer's buckets
   struct VariableLocator {
-    // Index into the `buckets_` variable.
+    // Index of the bucket containing the variable in the `buckets_` vector
     size_t bucket_index;
-    // Index of parameter in single bucket replica.
+    // Index of the variable in the bucket
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.hpp:445]: update this to mention it is the index of `variables`, `bucket_views_in`, etc structures?

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34399176).

**Review by zhaojuanmao — COMMENTED** [review]
nice work! the DDP reducer codes look much cleaner now, adding ci/master to make sure all the tests pass. Otherwise, codes look god to me!

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 3]
```diff
@@ -803,15 +797,12 @@ void Reducer::mark_variable_ready(size_t variable_index) {
   // TODO(@pietern): Make this work for both CPU/CUDA tensors.
   // When using CPU tensors we don't need to do this.
   // // Record event so that we can wait for all of them.
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:799]: (this is prior to this PR) duplicated "//"?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 4]
```diff
@@ -803,15 +797,12 @@ void Reducer::mark_variable_ready(size_t variable_index) {
...
-  // auto& event = replica.events[bucket_index.intra_bucket_index];
+  // auto& event = bucket.events[bucket_index.intra_bucket_index];
   // event.record();
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:801]: (can be done in followup PR) this comment seems still relevant. The autograd hook will fire on the same stream used by that function during forward, which might be different from the allreduce stream. I didn't we record that stream anywhere in reducer. 

https://github.com/pytorch/pytorch/blob/ec3a5ca6d3f7762922effc3d44132a7723bf8e16/torch/csrc/distributed/c10d/reducer.cpp#L376-L380

@zhaojuanmao @rohan-varma would I be correct that we should add this to guarantee correctness?

If we still need to make it work for CPU, today, we can use c10 events, which will work seamlessly for both CPU and GPU

Comment (mrshenli) [reply torch/csrc/distributed/c10d/reducer.cpp:801]: See how `ProcessGroupGloo` used `c10::event` and `c10::stream`. They basically become noop for CPU devices. 

https://github.com/pytorch/pytorch/blob/ec3a5ca6d3f7762922effc3d44132a7723bf8e16/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L330-L364

**Review by mrshenli — APPROVED** [review]
LGTM

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 4 above)
Comment (rohan-varma) [reply torch/csrc/distributed/c10d/reducer.cpp:801]: @mrshenli I did have a fix a while back for DDP under non-default stream: https://github.com/pytorch/pytorch/pull/40115/files

In this fix, we have a test where forward uses non-default stream and autograd engine restores these streams during backward. The fix for DDP was to sync with the current stream as opposed to default stream. Seems like that takes care of it, is there anything else we need to do?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 5]
```diff
@@ -803,15 +797,12 @@ void Reducer::mark_variable_ready(size_t variable_index) {
...
 
   // Check if this was the final gradient for this bucket.
-  if (--replica.pending == 0) {
-    // Kick off reduction if all replicas for this bucket are ready.
-    if (--bucket.pending == 0) {
-      mark_bucket_ready(bucket_index.bucket_index);
-    }
+  if (--bucket.pending == 0) {
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:804]: nice! makes the code simpler

**Review by rohan-varma — APPROVED** [review]
LGTM

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 4 above)
Comment (mrshenli) [reply torch/csrc/distributed/c10d/reducer.cpp:801]: Ah I see, yep, that should work. Thanks! In that case, we can remove Pieter's comments?

**Comment by awgu** [comment]
Re-landing since this did was not the breaking diff.

**Comment by awgu** [comment]
Making a brand new PR instead to deal with internal importing issues.

**Pull request closed without merging**